### ISSUE TITLE
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr! for git status

### ISSUE THREAD

**Issue by luminans** [git-for-windows/git#3655]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? 64-bit

```
$ git --version --build-options

git version 2.35.0.windows.1
```
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!

 - Which version of Windows are you running? Windows 10 64-bit

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19042.1466]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VisualStudioCode
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

 No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

CMD/PowerShell

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git status
```
 - What did you expect to occur after running these commands?

change status

 - What actually happened instead?

BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

** insert URL here **

**Comment by dscho** [comment]
> * What commands did you run to trigger this issue? If you can provide a
>       [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
>       this will help us understand the issue.
> 
> 
> ```
> git status
> ```

Hmm. I cannot reproduce.

Feel free to provide details, we can reopen the ticket easily enough.

**Comment by luminans** [comment]
I still have the repo having this issue locally after some changes.
What kind of information helps you reproduce the problem other than sharing the entire repo?

**Comment by dscho** [comment]
> What kind of information helps you reproduce the problem other than sharing the entire repo?

Have you tried following the guidance at https://stackoverflow.com/help/minimal-reproducible-example?

**Comment by benwitman** [comment]
FWIW I also got into this state, will try to see if I can get a repro, but in recent memory I fetched, then rebased a branch that resulted in a lot of conflicts, then git status returned this message. 
```
$>git --version
git version 2.35.0.vfs.0.0
```

**Comment by luminans** [comment]
Thanks. I have about the same memory.

**Comment by siprbaum** [comment]
I have run into exactly the same error message yesterday with
```
$ git version
git version 2.35.1.windows.1
```

At that time I was doing a `gh pr checkout <some_pull_request>`
The repo doesn't show any errors in `git fsck --full`. 
only some dangling commits and dangling blobs due to rebasing.

<details>
<summary>git gc made the problem go away</summary>

```
$ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!

siprbaum@notebook MINGW64 /d/src/git-tfs.backup ((2e204ccc...))
$ git gc
Enumerating objects: 17535, done.
Counting objects: 100% (17535/17535), done.
Delta compression using up to 12 threads
Compressing objects: 100% (3938/3938), done.
Writing objects: 100% (17535/17535), done.
Total 17535 (delta 13776), reused 17113 (delta 13454), pack-reused 0

siprbaum@notebook MINGW64 /d/src/git-tfs.backup ((2e204ccc...))
$ git status
Not currently on any branch.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   src/.paket/Paket.Restore.targets
        modified:   src/GitTfs.Vs2015/GitTfs.Vs2015.csproj
        modified:   src/GitTfs/Commands/Init.cs

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        Analysis-Globals.txt
        debugsettings.txt
        src/GitTfs.Vs2017/
        src/GitTfs.Vs2019/
        src/GitTfs.Vs2022/
        src/GitTfs.VsCommon/TfsHelper.Vs2017Base.cs

no changes added to commit (use "git add" and/or "git commit -a")
```
</details>

Trying to think of what `git gc` is doing and doing the steps manually to check after each one, I tried the following.

<details>
<summary>Running git pack-refs doesn't fix the problem, only `git pack-refs -all` does</summary>

```
$ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!
$ git pack-refs
$ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!
```


What helped was running 'git pack-refs --all`
```
$ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!
$ git pack-refs --all
$ git status
 <...  git status succeeded ... output skipped ...>
```
</details>

<details>
<summary>Narrowing it down to one ref which is also checked out as detached HEAD</summary>
  
reproducer script I used

```bash
#!/bin/sh

set -x
# Re-create pristine environment
rm -rf git-tfs.repro
cp -ar git-tfs.backup_cleaned git-tfs.repro

cd git-tfs.repro

# removing all packed-refs, and refs/* besides .git/refs/remotes/origin/master which is the one causing the problem
rm -rf .git/packed-refs 
find .git/refs -type f -not -path '.git/refs/remotes/origin/master' |xargs rm
find .git/refs -type d |xargs rmdir --ignore-fail-on-non-empty -p

#find .git/refs -type f |grep -v .git/refs/remotes/origin/master|xargs rm
#find .git/refs -type d |grep -v '.git/refs$'|xargs rmdir -p

# Show the remaining refs available
echo "#### Files/Folders in .git/refs ####" &&
find .git/refs/ &&

echo "#### git status ####"
git status >/dev/null ; echo $? # fails

mv .git/refs/remotes/origin/{master,renamed_master}
git status >/dev/null ; echo $? # succeeds, error is gone

mv .git/refs/remotes/origin/{renamed_master,master}
git status >/dev/null ; echo $? # fails again

# Show that HEAD is detached and points to the same commit as refs/remotes/origin/master
echo "#### Content of .git/HEAD ####"
cat .git/HEAD
od -c < .git/HEAD

echo "#### Content .git/refs/remotes/origin/master" &&
cat .git/refs/remotes/origin/master
od -c < .git/refs/remotes/origin/master
```

Output:
```
+ rm -rf git-tfs.repro
+ cp -ar git-tfs.backup_cleaned git-tfs.repro
+ cd git-tfs.repro
+ rm -rf .git/packed-refs
+ find .git/refs -type f -not -path .git/refs/remotes/origin/master
+ xargs rm
+ find .git/refs -type d
+ xargs rmdir --ignore-fail-on-non-empty -p
+ echo '#### Files/Folders in .git/refs ####'
#### Files/Folders in .git/refs ####
+ find .git/refs/
.git/refs/
.git/refs/remotes
.git/refs/remotes/origin
.git/refs/remotes/origin/master
+ echo '#### git status ####'
#### git status ####
+ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!
+ echo 3
3
+ mv .git/refs/remotes/origin/master .git/refs/remotes/origin/renamed_master
+ git status
+ echo 0
0
+ mv .git/refs/remotes/origin/renamed_master .git/refs/remotes/origin/master
+ git status
BUG: refs/files-backend.c:465: returning non-zero -1, should have set myerr!
+ echo 3
3
+ echo '#### Content of .git/HEAD ####'
#### Content of .git/HEAD ####
+ cat .git/HEAD
2e204cccef7207d79fae781c58847b6fe3eaaed3
+ od -c
0000000   2   e   2   0   4   c   c   c   e   f   7   2   0   7   d   7
0000020   9   f   a   e   7   8   1   c   5   8   8   4   7   b   6   f
0000040   e   3   e   a   a   e   d   3  \n
0000051
+ echo '#### Content .git/refs/remotes/origin/master'
#### Content .git/refs/remotes/origin/master
+ cat .git/refs/remotes/origin/master
2e204cccef7207d79fae781c58847b6fe3eaaed3
+ od -c
0000000   2   e   2   0   4   c   c   c   e   f   7   2   0   7   d   7
0000020   9   f   a   e   7   8   1   c   5   8   8   4   7   b   6   f
0000040   e   3   e   a   a   e   d   3  \n
0000051
```
</details>

To summarize, I have a remote reference where the commit it references is also checked out as detached HEAD.
If I rename that remote ref to some other name, the problem is gone. 
If I rename it back to its original name, the problem reappears.
